doc/viper/viper.md

```diff
@@ -0,0 +1,298 @@
+For example:
+
+```go
+package azblobbackupstorage
+
+import (
+	"github.com/spf13/pflag"
+
+	"vitess.io/vitess/go/viperutil"
+	"vitess.io/vitess/go/vt/servenv"
+)
+
+var (
+	configKey = viperutil.KeyPrefixFunc("backup.storage.azblob")
+
+	accountName = viperutil.Configure(
+		configKey("account.name"),
+		viperutil.Options[string]{
+			EnvVars:  []string{"VT_AZBLOB_ACCOUNT_NAME"},
+			FlagName: "azblob_backup_account_name",
+		},
+	)
+
+	accountKeyFile = viperutil.Configure(
+		configKey("account.key_file"),
+		viperutil.Options[string]{
+			FlagName: "azblob_backup_account_key_file",
+		},
+	)
+)
+
+func registerFlags(fs *pflag.FlagSet) {
+	fs.String("azblob_backup_account_name", accountName.Default(), "Azure Storage Account name for backups; if this flag is unset, the environment variable VT_AZBLOB_ACCOUNT_NAME will be used.")
+	fs.String("azblob_backup_account_key_file", accountKeyFile.Default(), "Path to a file containing the Azure Storage account key; if this flag is unset, the environment variable VT_AZBLOB_ACCOUNT_KEY will be used as the key itself (NOT a file path).")
+
+    viperutil.BindFlags(fs, accountName, accountKeyFile)
+}
+
+func init() {
+	servenv.OnParseFor("vtbackup", registerFlags)
+	servenv.OnParseFor("vtctl", registerFlags)
+	servenv.OnParseFor("vtctld", registerFlags)
+	servenv.OnParseFor("vttablet", registerFlags)
+}
+```
+
+## Config Files
+
+`viperutil` provides a few flags that allow binaries to read values from config files in addition to defaults, environment variables and flags.
+They are:
+
+- `--config-path`
+    - Default: `$(pwd)`
+    - EnvVar: `VT_CONFIG_PATH` (parsed exactly like a `$PATH` style shell variable).
+    - FlagType: `StringSlice`
+    - Behavior: Paths for `ReadInConfig` to search.
+- `--config-type`
+    - Default: `""`
+    - EnvVar: `VT_CONFIG_TYPE`
+    - FlagType: `flagutil.StringEnum`
+        - Values: everything contained in `viper.SupportedExts`, case-insensitive.
+    - Behavior: Force viper to use a particular unmarshalling strategy; required if the config file does not have an extension (by default, viper infers the config type from the file extension).
+- `--config-name`
+    - Default: `"vtconfig"`
+    - EnvVar: `VT_CONFIG_NAME`
+    - FlagType: `string`
+    - Behavior: Instructs `ReadInConfig` to only look in `ConfigPaths` for files named with this name (with any supported extension, unless `ConfigType` is also set, in which case only with that extension).
+- `--config-file`
+    - Default: `""`
+    - EnvVar: `VT_CONFIG_FILE`
+    - FlagType: `string`
+    - Behavior: Instructs `ReadInConfig` to search in `ConfigPaths` for explicitly a file with this name. Takes precedence over `ConfigName`.
+- `--config-file-not-found-handling`
+    - Default: `WarnOnConfigFileNotFound`
```

Comment:
I think this should be `exit`, not `warn`.

Reply:
this would break anyone moving to this version (because they wouldn't be able to specify a config file before, so they would immediately `exit` rather than use flags + default values, which was the behavior in the previous version)

Reply:
the way I understood this is that "IF a config file is specified AND it is not found, what do I do?". what you are saying is implying that this is simply "IF a config file is not found, what do I do?" We should make it mean the other thing. Is that possible?

Reply:
we _can_ (probably, we'd have to shim a few things and i haven't looked but it'll almost certainly work), but let's back up for a second first.

you don't "specify a config file" (usually, more on this in a bit); instead you specify enough for viper to _find_ a config file (by specifying any or all of: (1) paths to search; (2) the name of the file, without the extension; and (3) the type of config file, e.g. "yaml", or let viper infer it from the file extension) and viper will just find it. so the way we have this hooked up now is, "for free" if you run a binary while having a `vtconfig` file (in any supported format, so `.json`, `.yaml`, `.toml`, and so on (see [`SupportedExts`](https://github.com/spf13/viper/blob/v1.14.0/viper.go#L419))) in your working directory, the config gets loaded without you needing to add extra flags on the command line.

if we wanted the behavior you're describing, we'd have to remove that "for free" behavior (which may be fine and what we want, in the end, but that is the necessary tradeoff).

second, viper let's you explicitly specify a full path, including the extension, and it'll skip the searching process described above and just try to load that file. _this_ case i'm much more on board with what you're describing (though it would make the warn/error/exit flag more confusing, because in the _first_ case we probably want the default of "warn" but in _this_ case we probably want the default of "exit")

with all that said (and let me know if anything wasn't clear!!), uhh, for lack of a more eloquent question: thoughts?

Reply:
If a full path is explicitly specified, and the file is not found, exit.
If viper is searching various paths and no files are found, Info / Warn is sufficient.

Reply:
right, but what is the default value for this flag, then? you can't have a default value that's conditional on the final value of some other flag

Reply:
oh, actually i changed this to Ignore because it was breaking tests that invoked a `vtctlclient` and asserted on the output (since the warning now came first). let me try to actually fix that instead of being lazy, because i would prefer the Warn-as-default